be/src/runtime/load_channel.cpp

```diff
@@ -0,0 +1,163 @@
+            TabletsChannelKey key(params.id(), index_id);
+            channel.reset(new TabletsChannel(key, _mem_tracker.get()));
+            _tablets_channels.insert({index_id, channel});
+        }
+    }
+
+    RETURN_IF_ERROR(channel->open(params));
+
+    if (!_opened) {
+        _opened = true;
+    }
+    _last_updated_time = time(nullptr);
+    return Status::OK();
+}
+
+static void dummy_deleter(const CacheKey& key, void* value) {
+}
+
+Status LoadChannel::add_batch(
+        const PTabletWriterAddBatchRequest& request,
+        google::protobuf::RepeatedPtrField<PTabletInfo>* tablet_vec) {
+
+    int64_t index_id = request.index_id();
+    // 1. get tablets channel
+    std::shared_ptr<TabletsChannel> channel;
+    {
+        std::lock_guard<std::mutex> l(_lock);
+        auto it = _tablets_channels.find(index_id);
+        if (it == _tablets_channels.end()) {
+            auto handle = _lastest_success_channel->lookup(std::to_string(index_id));
+            // success only when eos be true
+            if (handle != nullptr && request.has_eos() && request.eos()) {
+                _lastest_success_channel->release(handle);
+                return Status::OK();
+            }
+            std::stringstream ss;
+            ss << "load channel " << _load_id << " add batch with unknown index id: " << index_id;
+            return Status::InternalError(ss.str());
+        }
+        channel = it->second;
+    }
+
+    // 2. add batch to tablets channel
+    if (request.has_row_batch()) {
+        RETURN_IF_ERROR(channel->add_batch(request));
+    }
+
+    // 3. check if mem consumption exceed limit
+    {
+        std::lock_guard<std::mutex> l(_lock);
+        if (_mem_tracker->limit_exceeded()) {
+            VLOG(1) << "mem consumption: " << _mem_tracker->consumption() << " exceed limit. load id: " << _load_id;
+            std::shared_ptr<TabletsChannel> channel;
+            if (_find_largest_max_consumption_tablets_channel(&channel)) {
+                channel->reduce_mem_usage();
+            } else {
+                // should not happen, add log to observe
+                LOG(WARNING) << "failed to find suitable tablets channel when mem limit execeed: " << _load_id;
+            }
+        }
+    }
+
+    // 4. handle eos
+    Status st;
+    if (request.has_eos() && request.eos()) {
+        bool finished = false;
+        RETURN_IF_ERROR(channel->close(request.sender_id(), &finished, request.partition_ids(), tablet_vec));
+        if (finished) {
+            std::lock_guard<std::mutex> l(_lock);
+            _tablets_channels.erase(index_id);
+            auto handle = _lastest_success_channel->insert(
+                    std::to_string(index_id), nullptr, 1, dummy_deleter);
+            _lastest_success_channel->release(handle);
+        }
+    }
+    _last_updated_time = time(nullptr);
+    return st;
+}
+
+bool LoadChannel::_find_largest_max_consumption_tablets_channel(std::shared_ptr<TabletsChannel>* channel) {
```

Comment:
add comment for lock should be held.
1 + 1 + 80 < 50 + 50
why not choose 80?

Reply:
It may happen but with low possibility. And even if we choose 50, it is not a big deal.
I will eave it unchanged. If it becomes a real case, change it later.